library/core/src/intrinsics.rs

```diff
@@ -784,6 +784,42 @@ extern "rust-intrinsic" {
     /// This intrinsic does not have a stable counterpart.
     pub fn breakpoint();
 
+    /// Selects which function to call depending on the context.
+    ///
+    /// If this function is evaluated at compile-time, then a call to this
+    /// intrinsic will be replaced with a call to `called_in_const`. It gets
+    /// replaced with a call to `called_at_rt` otherwise.
+    ///
+    /// # Type Requirements
+    ///
+    /// The two functions must be both function items. They cannot be function
+    /// pointers or closures.
+    ///
+    /// `arg` will be the arguments that will be passed to either one of the
+    /// two functions, therefore, both functions must accept the same type of
+    /// arguments. Both functions must return RET.
+    ///
+    /// # Safety
+    ///
+    /// This intrinsic allows breaking [referential transparency] in `const fn`
+    /// and is therefore `unsafe`.
+    ///
+    /// Code that uses this intrinsic must be extremely careful to ensure that
+    /// `const fn`s remain referentially-transparent independently of when they
+    /// are evaluated.
+    ///
+    /// The Rust compiler assumes that it is sound to replace a call to a `const
+    /// fn` with the result produced by evaluating it at compile-time. If
+    /// evaluating the function at run-time were to produce a different result,
+    /// or have any other observable side-effects, the behavior is undefined.
+    ///
+    /// [referential transparency]: https://en.wikipedia.org/wiki/Referential_transparency
+    #[cfg(not(bootstrap))]
+    // Although it is const stable, it should never be exposed outside core::intrinsics,
+    // and never stabilized.
+    #[rustc_const_stable(feature = "const_eval_select", since = "1.57")]
+    pub fn const_eval_select<ARG, F, G, RET>(arg: ARG, called_in_const: F, called_at_rt: G) -> RET;
```

Comment:
Do you think we can avoid some of the typeck complexity by adding Fn trait bounds to F and G?